src/lib/query/useBoundQueryFn.ts

```diff
@@ -0,0 +1,32 @@
+import { useDataEngine } from '@dhis2/app-runtime'
+import { useMemo } from 'react'
+import { QueryFunctionContext } from 'react-query'
+import type { DataEngine, Query, ResourceQuery } from '../../types'
+// types not exported from app-runtime...
+
+export const createBoundQueryFn =
+    (engine: DataEngine) =>
+    <TData>({ queryKey: [query], signal }: QueryFunctionContext<[Query]>) =>
+        engine.query(query, { signal }) as Promise<TData> // engine.query is not generic...
```

Comment:
What type-error are you seeing? Or do you mean `as Promise<TData>`. 
The problem is that `engine.query` returns a `JsonMap`-type that is just impossible to work with. So I just assert it to be `TData`, much like if `engine.query` would've been generic.

Reply:
mmm i think i am getting an error because Query is defined as Parameters<typeof useDataQuery>[0], and getting an index from a list means it might be undefined? But we can look at this together, nothing urgent.

Reply:
Do you still see the error after I refactored and extracted the querykeys to it's own type with `Readonly`?